chrome/content/zotero/xpcom/itemTreeManager.js

```diff
@@ -0,0 +1,276 @@
+/*
+	***** BEGIN LICENSE BLOCK *****
+
+	Copyright © 2019 Corporation for Digital Scholarship
+					 Vienna, Virginia, USA
+					 https://digitalscholar.org
+
+	This file is part of Zotero.
+
+	Zotero is free software: you can redistribute it and/or modify
+	it under the terms of the GNU Affero General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	Zotero is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU Affero General Public License for more details.
+
+	You should have received a copy of the GNU Affero General Public License
+	along with Zotero.  If not, see <http://www.gnu.org/licenses/>.
+
+	***** END LICENSE BLOCK *****
+*/
+
+const { COLUMNS: ITEMTREE_COLUMNS } = require("zotero/itemTreeColumns");
+
+/**
+ * @typedef {import("../itemTreeColumns.jsx").ItemTreeColumnOptions} ItemTreeColumnOptions
+ */
+
+class ItemTreeManager {
+	_observerAdded = false;
+	/** @type {ItemTreeColumnOptions[]} */
+	_customColumns = [];
+	/** 
+	 * Cached custom columns' data source, dataKey: dataProvider
+	 * @type {Record<string, (item: Zotero.Item, field: string, unformatted?: boolean) => string>}
+	 */
+	_customDataProvider = {};
+	constructor() {
+	}
+
+	/** 
+	 * Register a custom column. All registered columns must be valid, and must have a unique dataKey.
+	 * Although it's async, resolving does not promise the item trees are updated.
+	 * @param {ItemTreeColumnOptions | ItemTreeColumnOptions[]} options - An option or array of options to register
+	 * @returns {boolean} true if the column(s) are registered
+	 * @example
+	 * ```js
+	 * Zotero.ItemTreeManager.registerColumns(
+	 * {
+	 *     dataKey: 'rtitle',
+	 *     iconPath: 'chrome://zotero/skin/tick.png',
+	 *     label: 'reversed title',
+	 *     dataProvider: (item, field) => {
+	 *         return item.getField('title').split('').reverse().join('')}
```

Comment:
This isn't a good example, since it doesn't use the parameters provided. We can still have an example that uses those without having an API for custom fields.